cli/listener/src/main.rs

```diff
@@ -452,6 +453,32 @@ fn get_date_params<'a, I: IntoIterator<Item = (Cow<'a, str>, Cow<'a, str>)>>(
                 })?;
                 end_opt = Some(t);
             }
+            "days" => {
+                let days = v.parse::<i64>().map_err(|_| {
+                    ResponseError::BadRequest(
+                        "Invalid number of days in 'days' query parameter. \
+                    Expected e.g. '30'.",
+                    )
+                })?;
+                let end = chrono::Utc::now();
```

Comment:
If we call `now()` in this function, then this function will behave differently every time you call it: the return value will be different, even if the arguments are the same. This makes the function [impure](https://en.wikipedia.org/wiki/Pure_function). Impure functions are difficult to test, because their output is a moving target. Pure functions are much nicer to test, because you can just write assertions of the form `assert_eq(f(inputs), expected_output);`.

Aside from that, if we call `now()` directly, we may call it in multiple places and end up with inconsistencies. It’s not the case here, but imagine we had `begin_days_ago` and `end_days_ago` query params that set `begin` and `end` to `now().checked_sub_signed(begin_days_ago)` and `now().checked_sub_signed(end_days_ago)`. I would expect, if I pass `begin_days_ago=1&end_days_ago=0` that `end - begin` is a duration of exactly one day. But this is not the case! Between the first call to `now()`, and the second call to `now()`, some time elapsed, so we end up with a duration that is more than 1 day. Again, in this case it’s no big deal, but in general this can lead to very subtle bugs, so I prefer to avoid this. (An exception to this is if you want to measure how long some operation takes, but then you should use a monotonic clock and not `DateTime<Utc>`.)

The solution to both problems is to push the impurity to the call site. Instead of calling `now()` directly, add an argument `now: DateTime<Utc>` to the function, and let the caller deal with it.

* Now our function is pure, and therefore easy to test.
* If we reference `now` in multiple places, they will be consistent.

This only moved the problem though, but you can usually continue moving the problem, and then we will end up at a very natural place to call `now()`. In this case that would be the entry point of the request handler, `serve_request`.